### DISCUSSION TITLE
Dag that runs a GcsSensorOperator creates always 1 running instance and queues 16 in parallel when deploying the dag code.

### DISCUSSION THREAD

**Discussion by MaximM95** [apache/airflow#25682] [Q&A]
### Apache Airflow Provider(s)

google

### Versions of Apache Airflow Providers

apache-airflow-providers-google==5.1.0

### Apache Airflow version

2.1.4

### Operating System

Google Cloud Composer (GKE Autopilot, composer-2.0.22-airflow-2.1.4)

### Deployment

Composer

### Deployment details

Deployment via Terraform, see https://cloud.google.com/composer/docs/composer-2/create-environments#step_optional_select_a_service_account_for_your_environment

### What happened

We recently upgraded our Cloud Composer from version Composer-1.17.2-Airflow-2.1.2 to Composer-2.0.22-Airflow-2.1.4

When deploying a dag which runs a custom GcsBucketSensor that inherits from BaseSensorOperator, one instance runs and 16 others are queued. If I close one, another one is queued or If one sensor task has finished successful than another one will run and one further will be queued. It means that the scheduler is permanently trying to run 17 instances in parallel. 

That is the dag config.
dag = DAG(
     dag_id=Path(__file__).stem,
     schedule_interval=timedelta(seconds=1),
     start_date=days_ago(1),
     default_args={},
     max_active_runs=1,
     catchup=False

What could be the reason that the Composer Scheduler scheduled 17 DagRuns?

### What you think should happen instead

Only 1 active DagRun without permanently queueing 16 further DagRuns.

### How to reproduce

1. Create a Composer-2.0.22-Airflow-2.1.4 environment on GCP
2. Create a DAG that runs a sensor that inherits from BaseSensorOperator
3. Deploy/Upload the dag code to the /dags folder and the sensor to the /plugins folder in the Composer Bucket
4. The sensor will run automatically

### Anything else

In Composer v1 this problem has never occured. This problem does not destroy anything in the dag but it looks weird in the logs that always 16 further dags are scheduled.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by eladkal** [comment]
This sounds like something you need to ask/report to Composer support

**Comment by potiuk** [comment]
I wonder though - what kind of DagRuns are scheduled? Is it the same DagRun? other DagRun? What you see in the screen/logs? This might be side-effect of some errror you made in your dag (or indeed something wrong on the composer side) - hard to say without seeing more details.

One thing that could have caused that and something you should generally avoid is using `start_date=days_ago(1)` - this is not recommended practice (there used to be some examples in the past suggesting it but if you look at Airflow 2 recent docs, we converted all our examples to used fixed start_date. You should also put some fixed date there that will be the "real" date your DAG has started to run. You have catchup=False so it should not create more runs, but maybe it somehow triggers it.

**Comment by MaximM95** [comment]
Thank you for your answers!

This dag is a sensor that scans a GCS Bucket for files. Depending on the arriving file, it triggers further dags.

It seems that the Airflow scheduler tries to run 17 DagRuns of that sensor in parallel. Since it is only one active run allowed, 1 DagRun is running and the other 16 are queued. I do not see any errors in the logs. And the one running instance is also working fine and does what it has to do. The only problem is that it looks weird in the logs where 16 runs are scheduled. People will ask questions to me. I could also hide the logs when I do something like:

session.query(DagRun).filter(DagRun.state == State.QUEUED).filter(DagRun.dag_id.like("%sensor_dag%")).delete()

But I want to know what kind of problem the new Airflow instance has. Furthermore this does not stop the problem. The dags will be scheduled again and again. And what I don't quite understand is why this problem didn't appear in Composer v1 but it did in Composer v2 and why the Airflow scheduler is trying to run just 17 runs in parallel.

I have also tried to change the Start_Date to a fixed one but nothing has changed. And the catchup false should prevent the scheduling of multiple instances.

Either there is a bug in the Airflow Scheduler or it's a bug on Google's side. I'll definitely check with Google as well. But you are not aware of any such problems with the Airflow Scheduler?

**Reply by potiuk** [reply]
This is a strange pattern. BTW. Airflow 2.3 (already) and 2.4 (in a few weeks) will address your needs in a much better way - you will be able to run Data-driven scheduling (2.4) and dynamic task mapping (2.3) and both combined should enable single-DAG scenario for what you want to do .

And no - not aware of any problem like that.

I recommend you read and follow https://airflow.apache.org/docs/apache-airflow/stable/concepts/scheduler.html?highlight=performance#fine-tuning-your-scheduler-performance as well.
